yjit/src/codegen.rs

```diff
@@ -3971,6 +4002,191 @@ fn jit_thread_s_current(
+        ctx.stack_pop(2);
+        jit_putobject(jit, ctx, asm, result);
+        return true;
+    } else {
+        chain_guard_object_is_heap(
+            jit,
+            ctx,
+            asm,
+            ocb,
+            ctx.stack_opnd(0),
+            StackOpnd(0),
+            SEND_MAX_DEPTH,
+            side_exit
+            );
+
+        let arg = ctx.stack_pop(1);
+        let _recv = ctx.stack_pop(1);
+
+        let return_true  = asm.new_label("return_true");
+        let return_false = asm.new_label("return_false");
+
+        let arg_val = asm.load(arg);
+
+        // For heap objects we use the superclasses array to determine whether the given class is
+        // in the class heirarchy of the given object. From here this always finds us an exact
+        // result without additional chain guards or side exits.
+        // See class_search_class_ancestor in object.c
+
+        // First we check for an exact match
+        asm.comment("return true if cl == c");
+        let klass_opnd = asm.load(Opnd::mem(64, arg_val, RUBY_OFFSET_RBASIC_KLASS));
+        asm.cmp(klass_opnd, klass.into());
+        asm.je(return_true);
+
+        // Next we do a bounds check on the superclass array. If the given class depth is outside
+        // the object's class's depth, we know it is not inherited.
+        asm.comment("return false if cl.depth <= c.depth");
+        let klass_depth = unsafe { rb_RCLASS_SUPERCLASS_DEPTH(klass) };
+        let depth_opnd = Opnd::mem(64, klass_opnd, RUBY_OFFSET_RCLASS_SUPERCLASS_DEPTH);
+        asm.cmp(depth_opnd, klass_depth.into());
+        asm.jbe(return_false);
+
+        // Finally check for the class in the superclass array at the expected index.
+        asm.comment("return cl.superclasses[c.depth] == c");
+        let superclasses_opnd = asm.load(Opnd::mem(64, klass_opnd, RUBY_OFFSET_RCLASS_SUPERCLASSES));
+        let superclasses_off = Opnd::mem(64, superclasses_opnd, klass_depth * SIZEOF_VALUE as i32);
+        asm.cmp(superclasses_off, klass.into());
+        asm.jne(return_false);
+
+        // fall through to return true
+        let stack_ret = ctx.stack_push(Type::UnknownImm);
+
+        {
+            asm.write_label(return_true);
+
+            // Create a new context where the pushed value is of type true
+            let mut true_ctx = *ctx;
+            true_ctx.reset_chain_depth();
+            true_ctx.upgrade_opnd_type(StackOpnd(0), Type::True);
+
+            asm.mov(stack_ret, Qtrue.into());
+            let next_block = BlockId {
+                iseq: jit.iseq,
+                idx: jit_next_insn_idx(jit),
+            };
+            gen_branch(jit, ctx, asm, ocb, next_block, &true_ctx, None, None, gen_branch_direct);
+        }
+
+        {
+            asm.write_label(return_false);
+
+            ctx.upgrade_opnd_type(StackOpnd(0), Type::False);
+            asm.mov(stack_ret, Qfalse.into());
+
+            // fall through
+        }
```

Comment:
I wish this section was simpler. This gives us two labels, `return_true` and `return_false`, which pushes `Qtrue`/`Qfalse` while also setting that type information in the ctx (so that it can be used by `branchif`/`branchunless` without emitting any code). I'd like to think that in the future we can remove these labels and instead use the `ctx` to skip storing the value on the stack at all (ie. via the unused `ConstMapping` `TempMapping`).